// ngx-timeago/esm2015/timeago.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive, Input, ElementRef, Optional, ChangeDetectorRef } from '@angular/core';
import { Subject } from 'rxjs';
import { filter } from 'rxjs/operators';
import { TimeagoClock } from './timeago.clock';
import { TimeagoFormatter } from './timeago.formatter';
import { TimeagoIntl } from './timeago.intl';
import { isDefined, coerceBooleanProperty, dateParser } from './util';
export class TimeagoDirective {
    /**
     * @param {?} intl
     * @param {?} cd
     * @param {?} formatter
     * @param {?} element
     * @param {?} clock
     */
    constructor(intl, cd, formatter, element, clock) {
        this.cd = cd;
        this.clock = clock;
        /**
         * Emits on:
         * - Input change
         * - Intl change
         * - Clock tick
         */
        this.stateChanges = new Subject();
        this._live = true;
        if (intl) {
            this.intlSubscription = intl.changes.subscribe(() => this.stateChanges.next());
        }
        this.stateChanges.subscribe(() => {
            this.setContent(element.nativeElement, formatter.format(this.date));
            this.cd.markForCheck();
        });
    }
    /**
     * The Date to display. An actual Date object or something that can be fed to new Date.
     * @return {?}
     */
    get date() {
        return this._date;
    }
    /**
     * @param {?} date
     * @return {?}
     */
    set date(date) {
        this._date = dateParser(date).valueOf();
        if (this._date) {
            if (this.clockSubscription) {
                this.clockSubscription.unsubscribe();
                this.clockSubscription = undefined;
            }
            this.clockSubscription = this.clock.tick(date)
                .pipe(filter(() => this.live, this))
                .subscribe(() => this.stateChanges.next());
        }
        else {
            throw new SyntaxError(`Wrong parameter in TimeagoDirective. Expected a valid date, received: ${date}`);
        }
    }
    /**
     * If the directive should update itself over time
     * @return {?}
     */
    get live() {
        return this._live;
    }
    /**
     * @param {?} live
     * @return {?}
     */
    set live(live) {
        this._live = coerceBooleanProperty(live);
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        this.stateChanges.next();
    }
    /**
     * @param {?} node
     * @param {?} content
     * @return {?}
     */
    setContent(node, content) {
        if (isDefined(node.textContent)) {
            node.textContent = content;
        }
        else {
            node.data = content;
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.intlSubscription) {
            this.intlSubscription.unsubscribe();
            this.intlSubscription = undefined;
        }
        if (this.clockSubscription) {
            this.clockSubscription.unsubscribe();
            this.clockSubscription = undefined;
        }
        this.stateChanges.complete();
    }
}
TimeagoDirective.decorators = [
    { type: Directive, args: [{
                selector: '[timeago]',
                exportAs: 'timeago',
            },] },
];
/** @nocollapse */
TimeagoDirective.ctorParameters = () => [
    { type: TimeagoIntl, decorators: [{ type: Optional }] },
    { type: ChangeDetectorRef },
    { type: TimeagoFormatter },
    { type: ElementRef },
    { type: TimeagoClock }
];
TimeagoDirective.propDecorators = {
    date: [{ type: Input }],
    live: [{ type: Input }]
};
function TimeagoDirective_tsickle_Closure_declarations() {
    /** @type {?} */
    TimeagoDirective.prototype.intlSubscription;
    /** @type {?} */
    TimeagoDirective.prototype.clockSubscription;
    /**
     * Emits on:
     * - Input change
     * - Intl change
     * - Clock tick
     * @type {?}
     */
    TimeagoDirective.prototype.stateChanges;
    /** @type {?} */
    TimeagoDirective.prototype._date;
    /** @type {?} */
    TimeagoDirective.prototype._live;
    /** @type {?} */
    TimeagoDirective.prototype.cd;
    /** @type {?} */
    TimeagoDirective.prototype.clock;
}
